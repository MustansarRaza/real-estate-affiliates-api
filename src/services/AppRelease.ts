import { AppRelease } from "affiliates-api/models";
import { Op } from "sequelize";

export interface AppVariablesDTO {
  min_version: string;
}
export interface SettingsDTO {
  version: string;
  appVariables: AppVariablesDTO;
}
class AppReleaseService {
  /**
   * Returns mobile application settings
   */
  async get(appName: string): Promise<SettingsDTO | null> {
    const release = await AppRelease.findOne({
      attributes: ["version", "appVariables"],
      where: {
        appName: {
          [Op.eq]: appName,
        },
      },
    });

    return release ? this.serialize(release) : null;
  }
  /**
   * Serializes a queried {@see AppRelease} model instance into a DTO.
   */
  private serialize(release: AppRelease): SettingsDTO {
    return {
      version: release.version.toString(),
      appVariables: {
        min_version: release.appVariables!.min_version,
      },
    };
  }
}
export default AppReleaseService;
