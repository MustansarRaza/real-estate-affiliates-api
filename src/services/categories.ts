import { ListingStatusKey, PlatformKey } from "@pf/taxonomies";
import {
  Address,
  AgencyConfiguration,
  Category,
  Listing,
  ListingStatus,
  Location,
  Platform,
  popularCategoryKeyToIndex,
  Project,
  ProjectAffiliateVisibilityID,
  ProjectStatus,
} from "affiliates-api/models";
import { col, fn, Op } from "sequelize";

interface CategoryAvailabilityDTO {
  availableInCities: CategoryCityAvailabilityDTO[];
  availableInProjects: number[];
}

interface CategoryCityAvailabilityDTO {
  cityID: number;
  unitsCount: number;
}

interface CategoryDTO {
  id: number;
  name: string;
  nameSingular: string;
  orderIndex: number;
  slug: string;
  parentID: number | null;
  popularityIndex: number;
  availableInCities: CategoryCityAvailabilityDTO[];
  availableInProjects: number[];
}

/**
 * Provides access to categories.
 */
class CategoriesService {
  /**
   * Gets a list of all available categories.
   */
  async list(agencyID: number | null): Promise<CategoryDTO[]> {
    let platformId = null;

    if (agencyID) {
      const agencyConfig = await AgencyConfiguration.findOne({
        attributes: ["platformId"],
        where: { agencyId: { [Op.eq]: agencyID } },
        raw: true,
      });
      platformId = agencyConfig?.platformId;
    }
    if (!platformId) {
      const platform = await Platform.findOne({
        attributes: ["id"],
        where: { platformKey: { [Op.eq]: PlatformKey.ZAMEEN } },
        raw: true,
      });
      platformId = platform.id;
    }

    const categories = await Category.findAll({
      where: { platformId: { [Op.eq]: platformId } },
      order: [
        ["type_order", "ASC"],
        ["type_id", "ASC"],
      ],
    });

    const categoryCitiesAndProjectsMap = await this.getCategoryCitiesAndProjectsMap(agencyID);

    return categories.map((category: Category) => ({
      id: category.id,
      name: category.name,
      nameSingular: category.nameSingular,
      orderIndex: category.orderIndex,
      slug: category.slug,
      parentID: category.parentID,
      popularityIndex: popularCategoryKeyToIndex(category.key),
      availableInCities: categoryCitiesAndProjectsMap[category.id]?.availableInCities || [],
      availableInProjects: categoryCitiesAndProjectsMap[category.id]?.availableInProjects || [],
    }));
  }

  /**
   * Gets a map of cities and Projects against a category in which
   * that category is available.
   */
  private async getCategoryCitiesAndProjectsMap(
    agencyID: number | null
  ): Promise<{
    [propertTypeId: number]: CategoryAvailabilityDTO;
  }> {
    const categoriesAvailablityResult = await Listing.findAll({
      attributes: [[fn("COUNT", "id"), "Count"]],
      include: [
        {
          model: ListingStatus,
          attributes: [],
          where: {
            key: {
              [Op.eq]: ListingStatusKey.AVAILABLE,
            },
          },
          required: true,
        },
        {
          model: Category,
          attributes: ["id"],
          required: true,
        },
        {
          model: Project,
          required: true,
          attributes: ["id"],
          where: {
            status: {
              [Op.eq]: ProjectStatus.ON,
            },
            affiliateVisibility: {
              [Op.eq]: ProjectAffiliateVisibilityID.VISIBLE,
            },
            ...(agencyID && {
              agencyID: {
                [Op.eq]: agencyID,
              },
            }),
          },
          include: [
            {
              model: Address,
              required: true,
              attributes: [],
              include: [
                {
                  model: Location,
                  attributes: ["id"],
                  required: true,
                },
              ],
            },
          ],
        },
      ],
      group: [
        col("Category.type_id"),
        col("Project.Address.Location.title"),
        col("Project.project_id"),
      ],
      raw: true,
    });

    const categoryCitiesAndProjectsMap: { [propertTypeId: number]: CategoryAvailabilityDTO } = {};
    categoriesAvailablityResult.forEach((row: any) => {
      const propertyTypeId = row["Category.id"];
      const projectID = row["Project.id"];
      const cityID = row["Project.Address.Location.id"];
      const unitsCount = row["Count"];

      if (categoryCitiesAndProjectsMap.hasOwnProperty(propertyTypeId)) {
        categoryCitiesAndProjectsMap[propertyTypeId].availableInProjects.push(projectID);
        const index = categoryCitiesAndProjectsMap[propertyTypeId].availableInCities
          .map((city: CategoryCityAvailabilityDTO) => city.cityID)
          .indexOf(cityID);
        if (index === -1) {
          categoryCitiesAndProjectsMap[propertyTypeId].availableInCities.push({
            cityID,
            unitsCount,
          });
        } else {
          categoryCitiesAndProjectsMap[propertyTypeId].availableInCities[
            index
          ].unitsCount += unitsCount;
        }
      } else {
        categoryCitiesAndProjectsMap[propertyTypeId] = {
          availableInProjects: [projectID],
          availableInCities: [{ cityID, unitsCount }],
        };
      }
    });

    return categoryCitiesAndProjectsMap;
  }
}

export default CategoriesService;
