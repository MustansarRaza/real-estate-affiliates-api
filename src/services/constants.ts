export enum ErrorMessages {
  USER_AFFILIATE_ERROR_MESSAGE = "Current user is not an affiliate",
}

export const FORMAT_DATE = {
  FULL_DATE_24_HOUR_FORMAT_DATE_FNS: "YYYY-MM-DD HH:mm:ss",
};
