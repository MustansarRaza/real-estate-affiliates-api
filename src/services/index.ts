import AppRelease from "./AppRelease";
import CategoriesService from "./categories";
import ClientsService, { CreateClientDTO as CreateClientDTOType } from "./clients";
import CommissionsService from "./commissions";
import LeadsService, { CreateLeadDTO as CreateLeadDTOType } from "./leads";
import ListingsService, { ListingDTO as ListingDTOType } from "./listings";
import LocationsService from "./locations";
import MeService, { CreateMeDTO as CreateMeDTOType, MeDTO as MeDTOType } from "./me";
import ProfileService from "./profile";
import ProjectsService from "./projects";
import ListingReservationService, { ListingReservationOptions } from "./reservations";
import TasksService from "./tasks";
import TaskTypesService from "./taskTypes";

export interface ServiceMesh {
  profile: ProfileService;
  me: MeService;
  clients: ClientsService;
  tasks: TasksService;
  taskTypes: TaskTypesService;
  leads: LeadsService;
  projects: ProjectsService;
  listings: ListingsService;
  locations: LocationsService;
  categories: CategoriesService;
  reservations: ListingReservationService;
  commissions: CommissionsService;
  appRelease: AppRelease;
}

const useServices = (): ServiceMesh => {
  const locations = new LocationsService();
  const profile = new ProfileService({ locations });
  const me = new MeService({ profile, locations });
  const clients = new ClientsService({ me, locations });
  const taskTypes = new TaskTypesService();
  const leads = new LeadsService({ me, taskTypes });
  const tasks = new TasksService({ leads, taskTypes });
  const listings = new ListingsService({ me });
  const commissions = new CommissionsService();
  const projects = new ProjectsService({ me, listings, commissions, locations });
  const categories = new CategoriesService();
  const reservations = new ListingReservationService({ me, listings });
  const appRelease = new AppRelease();

  return {
    profile,
    me,
    clients,
    leads,
    tasks,
    taskTypes,
    projects,
    listings,
    locations,
    categories,
    reservations,
    commissions,
    appRelease,
  };
};

export {
  useServices,
  ClientsService,
  LeadsService,
  MeService,
  ProfileService,
  ProjectsService,
  ListingsService,
  TasksService,
  TaskTypesService,
  LocationsService,
  CategoriesService,
  ListingReservationService,
  CommissionsService,
  AppRelease,
};

export type CreateClientDTO = CreateClientDTOType;
export type CreateLeadDTO = CreateLeadDTOType;
export type CreateMeDTO = CreateMeDTOType;
export type CreateReservationDTO = ListingReservationOptions;
export type ListingDTO = ListingDTOType;
export type MeDTO = MeDTOType;
