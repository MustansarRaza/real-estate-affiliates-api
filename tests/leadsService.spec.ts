import { SourcePlatform } from "@pf/taxonomies";
import {
  Address,
  Lead,
  LeadClassification,
  LeadClassificationID,
  LeadInquiryBudgetCurrency,
  LeadInquiryObjectType,
  LeadLocation,
  LeadLocationType,
  LeadMarket,
  LeadSide,
  LeadSourceMapping,
  LeadState,
  LeadStatus,
  Purpose,
  PurposeID,
} from "affiliates-api/models";
import { LeadsService } from "affiliates-api/services";
import { LeadDTO } from "affiliates-api/services/leads";
import MockDate from "mockdate";
import { Op } from "sequelize";
import { loadSequelizeFixtures, useMockedServices } from "./support";
import { createLead, ModelFixtureConstants, ModelFixtureNames } from "./support/data";

describe("LeadsService", () => {
  const mockedServices = useMockedServices();
  const service = new LeadsService(mockedServices);

  beforeEach(async () => {
    jest.resetAllMocks();
    mockedServices.me.owner.mockReturnValue(
      Promise.resolve({
        affiliateID: ModelFixtureConstants.verifiedAffiliateID,
        userID: ModelFixtureConstants.staffUserBDMID,
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      })
    );

    await loadSequelizeFixtures(
      ModelFixtureNames.AGENTS,
      ModelFixtureNames.AGENCY_CONFIGURATION,
      ModelFixtureNames.USERS,
      ModelFixtureNames.CONTACTS,
      ModelFixtureNames.CATEGORIES,
      ModelFixtureNames.DESIGNATIONS,
      ModelFixtureNames.COUNTRIES,
      ModelFixtureNames.LOCATIONS,
      ModelFixtureNames.AFFILIATES,
      ModelFixtureNames.AFFILIATE_CONTACTS,
      ModelFixtureNames.AFFILIATE_ASSIGNEES,
      ModelFixtureNames.POTENTIAL_AFFILIATE_VERIFICATION_STATUS,
      ModelFixtureNames.POTENTIAL_AFFILIATES,
      ModelFixtureNames.CLIENTS,
      ModelFixtureNames.LEAD_SOURCE,
      ModelFixtureNames.LEAD_STATUSES
    );
  });

  describe("list", () => {
    it("only returns active leads", async () => {
      const activeLead = await createLead(
        ModelFixtureConstants.staffUserBDMID,
        ModelFixtureConstants.activeAffiliateClientID,
        {
          state: LeadState.ON,
        }
      );
      await createLead(
        ModelFixtureConstants.staffUserBDMID,
        ModelFixtureConstants.activeAffiliateClientID,
        { state: LeadState.DELETED }
      );
      await createLead(
        ModelFixtureConstants.staffUserBDMID,
        ModelFixtureConstants.activeAffiliateClientID,
        { state: LeadState.PENDING }
      );

      const leads = await service.list({ ssoID: ModelFixtureConstants.verifiedSSOID });

      expect(leads).toHaveLength(1);
      expect(leads[0].id).toBe(activeLead.id);
    });

    it("should not return locationIds if leadLocationsRequired is set to false", async () => {
      await loadSequelizeFixtures(ModelFixtureNames.LEADS, ModelFixtureNames.LEAD_LOCATIONS);

      const leads = await service.list({ ssoID: ModelFixtureConstants.verifiedSSOID }, false);

      leads.forEach((lead: LeadDTO) => {
        expect(lead.inquiry.locationIDs.length).toEqual(0);
      });
    });

    [
      {},
      { fromDate: "2020-01-01T00:00:00Z" },
      { toDate: "2021-04-15T00:00:00Z" },
      { fromDate: "2020-01-01T00:00:00Z", toDate: "2020-12-31T00:00:00Z" },
      { fromDate: "2021-01-01T00:00:00Z", toDate: "2021-04-15T00:00:00Z" },
    ].forEach((dateRange) => {
      it(`should consider only the tasks that were created againt the lead in the provided interval i.e. fromDate : ${dateRange.fromDate}, toDate: ${dateRange.toDate}`, async () => {
        await loadSequelizeFixtures(
          ModelFixtureNames.LEADS,
          ModelFixtureNames.LEAD_LOCATIONS,
          ModelFixtureNames.TASK_TYPES,
          ModelFixtureNames.TASKS
        );

        //@ts-ignore
        mockedServices.taskTypes.getMappings.mockResolvedValue({
          "3": { id: 3, title: "Meeting (Arrange)", parentID: 62 },
          "6": { id: 6, title: "Token Received", parentID: 56 },
          "100": { id: 100, title: "Generate Token", parentID: 56 },
          "85": { id: 85, title: "Partial Down Payment", parentID: 56 },
          "52": { id: 85, title: "Closed (Won)", parentID: 56 },
        });

        const leads = await service.list(
          { ssoID: ModelFixtureConstants.verifiedSSOID },
          false,
          dateRange
        );

        leads.forEach((lead: LeadDTO) => {
          if (lead.lastTask) {
            if (dateRange.fromDate)
              expect(lead.lastTask.dateAdded >= dateRange.fromDate).toBeTruthy();
            if (dateRange.toDate) expect(lead.lastTask.dateAdded <= dateRange.toDate).toBeTruthy();
          }
        });
      });
    });

    it("sets purpose to for-sale when market is primary and there is no purpose", async () => {
      const primaryMarketLead = await createLead(
        ModelFixtureConstants.staffUserBDMID,
        ModelFixtureConstants.activeAffiliateClientID,
        {
          market: LeadMarket.PRIMARY,
          inquiryPurposeID: null,
        }
      );
      const leads = await service.list({ ssoID: ModelFixtureConstants.verifiedSSOID });

      expect(leads).toHaveLength(1);
      expect(leads[0].id).toBe(primaryMarketLead.id);
      expect(leads[0].inquiry.purpose).toBe(Purpose.FOR_SALE);
    });

    it("overwrites the purpose to for-sale when the market is primary", async () => {
      const primaryMarketLead = await createLead(
        ModelFixtureConstants.staffUserBDMID,
        ModelFixtureConstants.activeAffiliateClientID,
        {
          market: LeadMarket.PRIMARY,
          inquiryPurposeID: PurposeID.FOR_SALE as number,
        }
      );
      const leads = await service.list({ ssoID: ModelFixtureConstants.verifiedSSOID });

      expect(leads).toHaveLength(1);
      expect(leads[0].id).toBe(primaryMarketLead.id);
      expect(leads[0].inquiry.purpose).toBe(Purpose.FOR_SALE);
    });

    it("does not overwrite the purpose when the market is secondary", async () => {
      const secondaryMarketLead = await createLead(
        ModelFixtureConstants.staffUserBDMID,
        ModelFixtureConstants.activeAffiliateClientID,
        {
          market: LeadMarket.SECONDARY,
          inquiryPurposeID: PurposeID.FOR_RENT as number,
        }
      );
      const leads = await service.list({ ssoID: ModelFixtureConstants.verifiedSSOID });

      expect(leads).toHaveLength(1);
      expect(leads[0].id).toBe(secondaryMarketLead.id);
      expect(leads[0].inquiry.purpose).toBe(Purpose.FOR_RENT);
    });
  });

  describe("listSummary", () => {
    it("should return default summary if no fromDate is provided as well as there doesn't exist any lead against the affiliate", async () => {
      const defaultSummary = {
        active: 0,
        inactive: 0,
        tokenGenerated: 0,
        closedWon: 0,
        downPayment: 0,
        completeDownPayment: 0,
      };

      const leadsSummary = await service.listSummary(
        { ssoID: ModelFixtureConstants.verifiedSSOID },
        {}
      );

      expect(leadsSummary).toEqual(defaultSummary);
    });

    it("should return default summary if there doesn't exist any lead against the affiliate", async () => {
      const defaultSummary = {
        active: 0,
        inactive: 0,
        tokenGenerated: 0,
        closedWon: 0,
        downPayment: 0,
        completeDownPayment: 0,
      };

      const leadsSummary = await service.listSummary(
        { ssoID: ModelFixtureConstants.verifiedSSOID },
        { fromDate: "2020-01-01T00:00:00Z", toDate: "2020-12-31T00:00:00Z" }
      );

      expect(leadsSummary).toEqual(defaultSummary);
    });

    it("should return the summary having the inactive count immediately if there doesn't exist any task against the lead(i.e. no active lead)", async () => {
      await loadSequelizeFixtures(ModelFixtureNames.LEADS, ModelFixtureNames.LEAD_LOCATIONS);

      const defaultSummary = {
        active: 0,
        inactive: 4,
        tokenGenerated: 0,
        closedWon: 0,
        downPayment: 0,
        completeDownPayment: 0,
      };

      const leadsSummary = await service.listSummary(
        { ssoID: ModelFixtureConstants.verifiedSSOID },
        { fromDate: "2020-01-01T00:00:00Z", toDate: "2020-04-15T00:00:00Z" }
      );

      expect(leadsSummary).toEqual(defaultSummary);
    });

    it("should return the summary having the inactive count immediately if there doesn't exist any task against the lead(i.e. no active lead) and is created before the provided toDate in the date range", async () => {
      await loadSequelizeFixtures(ModelFixtureNames.LEADS, ModelFixtureNames.LEAD_LOCATIONS);

      const defaultSummary = {
        active: 0,
        inactive: 3,
        tokenGenerated: 0,
        closedWon: 0,
        downPayment: 0,
        completeDownPayment: 0,
      };

      const leadsSummary = await service.listSummary(
        { ssoID: ModelFixtureConstants.verifiedSSOID },
        { fromDate: "2020-03-01T00:00:00Z", toDate: "2020-03-12T00:00:00Z" }
      );

      expect(leadsSummary).toEqual(defaultSummary);
    });

    it("should return the default summary immediately if there doesn't exist any lead that is created before the provided toDate in the date range", async () => {
      await loadSequelizeFixtures(ModelFixtureNames.LEADS, ModelFixtureNames.LEAD_LOCATIONS);

      const defaultSummary = {
        active: 0,
        inactive: 0,
        tokenGenerated: 0,
        closedWon: 0,
        downPayment: 0,
        completeDownPayment: 0,
      };

      const leadsSummary = await service.listSummary(
        { ssoID: ModelFixtureConstants.verifiedSSOID },
        { fromDate: "2020-01-01T00:00:00Z", toDate: "2020-03-01T00:00:00Z" }
      );

      expect(leadsSummary).toEqual(defaultSummary);
    });

    [
      {},
      { fromDate: "2020-01-01T00:00:00Z" },
      { toDate: "2021-04-15T00:00:00Z" },
      { fromDate: "2020-01-01T00:00:00Z", toDate: "2020-12-31T00:00:00Z" },
      { fromDate: "2021-01-01T00:00:00Z", toDate: "2021-02-15T00:00:00Z" },
      { fromDate: "2021-02-15T00:00:00Z", toDate: "2021-03-31T00:00:00Z" },
      { fromDate: "2021-01-01T00:00:00Z", toDate: "2021-04-15T00:00:00Z" },
    ].forEach((dateRange) => {
      it(`should return the summary by considering only the tasks as well as winning tasks that were created against the lead in the provided interval i.e. fromDate : ${dateRange.fromDate}, toDate: ${dateRange.toDate}`, async () => {
        await loadSequelizeFixtures(
          ModelFixtureNames.LEADS,
          ModelFixtureNames.LEAD_LOCATIONS,
          ModelFixtureNames.TASK_TYPES,
          ModelFixtureNames.TASKS
        );

        //@ts-ignore
        mockedServices.taskTypes.getMappings.mockResolvedValue({
          "3": { id: 3, title: "Meeting (Arrange)", parentID: 53 },
          "6": { id: 6, title: "Token Received", parentID: 56 },
          "100": { id: 100, title: "Generate Token", parentID: 56 },
          "85": { id: 85, title: "Partial Down Payment", parentID: 56 },
          "52": { id: 85, title: "Closed (Won)", parentID: 56 },
        });

        const leadsSummary = await service.listSummary(
          { ssoID: ModelFixtureConstants.verifiedSSOID },
          dateRange
        );

        expect(leadsSummary).toMatchSnapshot();
      });
    });
  });

  describe("create", () => {
    const createLeadBaseDTO = {
      clientID: ModelFixtureConstants.activeAffiliateClientID,
      side: LeadSide.DEMAND,
      market: LeadMarket.PRIMARY,
      classification: LeadClassification.MODERATE,
      inquiry: {
        object: {
          type: LeadInquiryObjectType.PROJECT,
          id: 1,
        },
        purpose: Purpose.FOR_SALE,
        categoryID: 1,
        locationIDs: [3],
        budget: {
          amount: 5000,
          currency: LeadInquiryBudgetCurrency.EUR,
        },
        bedroomCount: 3,
        bathroomCount: 2,
        unitNumber: null,
      },
      source: SourcePlatform.AFFILIATES_WEB_APP,
    };

    // eslint-disable-next-line unicorn/consistent-function-scoping
    const queryLead = (leadID: number): Promise<Lead | null> =>
      Lead.findOne({
        where: {
          id: {
            [Op.eq]: leadID,
          },
        },
        include: [
          {
            model: LeadLocation,
          },
        ],
      });

    const queryLeadWithAddress = (leadID: number): Promise<Lead | null> =>
      Lead.findOne({
        where: {
          id: {
            [Op.eq]: leadID,
          },
        },
        include: [
          {
            model: LeadLocation,
          },
          {
            model: Address,
          },
        ],
      });

    const queryLeadSourceMapping = (leadId: number): Promise<LeadSourceMapping | null> =>
      LeadSourceMapping.findOne({
        where: {
          inquiryId: { [Op.eq]: leadId },
        },
      });
    const leadServiceMock: any = {
      mockShareWithSupport: jest.fn(),
    };

    beforeAll(() => {
      MockDate.set("2020-03-28T14:12:00.270Z");
    });

    afterAll(() => {
      MockDate.reset();
    });

    it("can create a lead", async () => {
      const leadDTO = await service.create(
        { ssoID: ModelFixtureConstants.verifiedSSOID },
        createLeadBaseDTO,
        leadServiceMock,
        {}
      );

      const leadSourceMappings = await queryLeadSourceMapping(leadDTO.id);

      const lead = await queryLead(leadDTO.id);
      expect({ ...lead?.toJSON(), id: null }).toMatchSnapshot();

      expect(lead).toBeTruthy();
      expect(lead?.id).toBe(leadDTO.id);
      expect(lead?.source).toBe(SourcePlatform.AFFILIATES_WEB_APP);
      expect(lead?.userID).toBe(ModelFixtureConstants.staffUserBDMID);
      expect(lead?.clientID).toBe(1);
      expect(lead?.statusID).toBe(ModelFixtureConstants.leadStatusNewID);
      expect(lead?.side).toBe(LeadSide.DEMAND);
      expect(lead?.market).toBe(LeadMarket.PRIMARY);
      expect(lead?.classificationID).toBe(LeadClassificationID.MODERATE);
      expect(lead?.inquiryObjectType).toBe("project");
      expect(lead?.inquiryObjectID).toBe(1);
      expect(lead?.inquiryPurposeID).toBe(PurposeID.FOR_SALE);
      expect(lead?.inquiryCategoryID).toBe(1);
      expect(lead?.inquiryBudgetAmount).toBe(5000);
      expect(lead?.inquiryBudgetCurrency).toBe("EUR");
      expect(lead?.inquiryBedroomCount).toBe(3);
      expect(lead?.inquiryBathroomCount).toBe(2);
      expect(lead?.LeadLocations).toHaveLength(1);
      expect(lead?.LeadLocations[0].leadID).toBe(lead?.id);
      expect(lead?.LeadLocations[0].locationID).toBe(3);
      expect(lead?.LeadLocations[0].type).toBe(LeadLocationType.INTERESTED_IN);
      expect(lead?.LeadLocations[0].createdAt).toBeTruthy();
      expect(lead?.createdAt).toBeTruthy();
      expect(lead?.updatedAt).toBeTruthy();

      expect(leadSourceMappings).toBeTruthy();
      expect(leadSourceMappings?.inquirySourceId).toBe(ModelFixtureConstants.affiliateLeadSourceId);
    });

    it("can create a lead not linked to an object", async () => {
      const leadDTO = await service.create(
        { ssoID: ModelFixtureConstants.verifiedSSOID },
        { ...createLeadBaseDTO, inquiry: { ...createLeadBaseDTO.inquiry, object: null } },
        leadServiceMock,
        {}
      );

      expect(leadDTO.inquiry.object).toBe(null);

      const lead = await queryLead(leadDTO.id);
      expect(lead?.inquiryObjectType).toBe(null);
      expect(lead?.inquiryObjectID).toBe(null);
    });

    it("can create a lead without any locations", async () => {
      const leadDTO = await service.create(
        { ssoID: ModelFixtureConstants.verifiedSSOID },
        {
          ...createLeadBaseDTO,
          inquiry: { ...createLeadBaseDTO.inquiry, locationIDs: [] },
        },
        leadServiceMock,
        {}
      );

      expect(leadDTO.inquiry.locationIDs).toEqual([]);

      const lead = await queryLead(leadDTO.id);
      expect(lead?.LeadLocations).toHaveLength(0);
    });

    it("does not allow creating a lead without having an owner", async () => {
      mockedServices.me.owner.mockReturnValue(Promise.resolve(null));

      await expect(
        service.create(
          { ssoID: ModelFixtureConstants.unverifiedSSOID },
          createLeadBaseDTO,
          leadServiceMock,
          {}
        )
      ).rejects.toMatchSnapshot();
    });

    it("does not allow creating a lead with an invalid category ID", async () => {
      await expect(
        service.create(
          { ssoID: ModelFixtureConstants.unverifiedSSOID },
          {
            ...createLeadBaseDTO,
            inquiry: { ...createLeadBaseDTO.inquiry, categoryID: 999 },
          },
          leadServiceMock,
          {}
        )
      ).rejects.toMatchSnapshot();
    });

    it("does not allow creating a lead with an invalid location ID", async () => {
      await expect(
        service.create(
          { ssoID: ModelFixtureConstants.unverifiedSSOID },
          {
            ...createLeadBaseDTO,
            inquiry: { ...createLeadBaseDTO.inquiry, locationIDs: [999] },
          },
          leadServiceMock,
          {}
        )
      ).rejects.toMatchSnapshot();
    });

    it("returns a nice error if the initial status is not available", async () => {
      LeadStatus.destroy({ truncate: true });

      await expect(
        service.create(
          { ssoID: ModelFixtureConstants.unverifiedSSOID },
          createLeadBaseDTO,
          leadServiceMock,
          {}
        )
      ).rejects.toMatchSnapshot();
    });

    it("properly creates a secondary market supply lead with unit number", async () => {
      const testUnitNumber = "4C-1234";
      const testLocationID = 3;
      const leadDTO = await service.create(
        { ssoID: ModelFixtureConstants.verifiedSSOID },
        {
          ...createLeadBaseDTO,
          market: LeadMarket.SECONDARY,
          side: LeadSide.SUPPLY,
          inquiry: {
            ...createLeadBaseDTO.inquiry,
            unitNumber: testUnitNumber,
            locationIDs: [testLocationID],
          },
        },
        leadServiceMock,
        {}
      );

      const lead = await queryLeadWithAddress(leadDTO.id);

      expect(lead).toBeDefined();
      expect(lead?.Address).toBeDefined();

      expect(lead?.Address?.locationID).toEqual(testLocationID);
      expect(lead?.Address?.unitNumber).toStrictEqual(testUnitNumber);
    });

    it("picks the status marked as the initial status", async () => {
      LeadStatus.destroy({ truncate: true });
      const [_, initialStatus] = await Promise.all([
        LeadStatus.create({
          title: "Not new",
          isInitial: false,
        }),
        LeadStatus.create({
          title: "New",
          isInitial: true,
        }),
      ]);

      const leadDTO = await service.create(
        { ssoID: ModelFixtureConstants.verifiedSSOID },
        createLeadBaseDTO,
        leadServiceMock,
        {}
      );

      const lead: Lead | null = await Lead.findOne({
        where: {
          id: {
            [Op.eq]: leadDTO.id,
          },
        },
      });

      expect(lead?.statusID).toBe(initialStatus.id);
    });
  });
});
