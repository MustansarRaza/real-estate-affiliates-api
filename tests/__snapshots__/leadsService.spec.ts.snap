// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeadsService create can create a lead 1`] = `
Object {
  "LeadLocations": Array [
    Object {
      "createdAt": 2020-03-28T14:12:00.000Z,
      "id": 1,
      "leadID": 1,
      "locationID": 3,
      "type": "interested_in",
    },
  ],
  "addressID": null,
  "agencyID": 1337,
  "classificationID": 3,
  "clientID": 1,
  "createdAt": 2020-03-28T14:12:00.000Z,
  "id": null,
  "inquiryBathroomCount": 2,
  "inquiryBedroomCount": 3,
  "inquiryBudgetAmount": 5000,
  "inquiryBudgetCurrency": "EUR",
  "inquiryCategoryID": 1,
  "inquiryObjectID": 1,
  "inquiryObjectType": "project",
  "inquiryPurposeID": 1,
  "market": "primary",
  "side": "demand",
  "source": "Propforce.com",
  "state": "on",
  "statusID": 1,
  "tsrClassificationID": 3,
  "updatedAt": 2020-03-28T14:12:00.000Z,
  "userID": 1,
}
`;

exports[`LeadsService create does not allow creating a lead with an invalid category ID 1`] = `
Object {
  "extra": Object {
    "ssoID": "b8d3b78f-d7c0-45f9-be04-1e168ae959d3",
  },
  "message": "Invalid category or location ID",
  "name": "BadRequest",
  "originalError": Object {
    "code": undefined,
    "message": "SQLITE_CONSTRAINT: FOREIGN KEY constraint failed",
    "name": "SequelizeForeignKeyConstraintError",
  },
  "statusCode": 400,
}
`;

exports[`LeadsService create does not allow creating a lead with an invalid location ID 1`] = `
Object {
  "extra": Object {
    "ssoID": "b8d3b78f-d7c0-45f9-be04-1e168ae959d3",
  },
  "message": "Invalid category or location ID",
  "name": "BadRequest",
  "originalError": Object {
    "code": undefined,
    "message": "SQLITE_CONSTRAINT: FOREIGN KEY constraint failed",
    "name": "SequelizeForeignKeyConstraintError",
  },
  "statusCode": 400,
}
`;

exports[`LeadsService create does not allow creating a lead without having an owner 1`] = `
Object {
  "extra": Object {
    "ssoID": "b8d3b78f-d7c0-45f9-be04-1e168ae959d3",
  },
  "message": "Current user has no owner",
  "name": "BadRequest",
  "originalError": undefined,
  "statusCode": 400,
}
`;

exports[`LeadsService create returns a nice error if the initial status is not available 1`] = `
Object {
  "extra": Object {
    "ssoID": "b8d3b78f-d7c0-45f9-be04-1e168ae959d3",
  },
  "message": "No initial status available",
  "name": "WrongSetup",
  "originalError": undefined,
  "statusCode": 500,
}
`;

exports[`LeadsService listSummary should return the summary by considering only the tasks as well as winning tasks that were created against the lead in the provided interval i.e. fromDate : 2020-01-01T00:00:00Z, toDate: 2020-12-31T00:00:00Z 1`] = `
Object {
  "active": 2,
  "closedWon": 0,
  "completeDownPayment": 0,
  "downPayment": 0,
  "inactive": 2,
  "tokenGenerated": 0,
}
`;

exports[`LeadsService listSummary should return the summary by considering only the tasks as well as winning tasks that were created against the lead in the provided interval i.e. fromDate : 2020-01-01T00:00:00Z, toDate: undefined 1`] = `
Object {
  "active": 3,
  "closedWon": 1,
  "completeDownPayment": 0,
  "downPayment": 0,
  "inactive": 1,
  "tokenGenerated": 0,
}
`;

exports[`LeadsService listSummary should return the summary by considering only the tasks as well as winning tasks that were created against the lead in the provided interval i.e. fromDate : 2021-01-01T00:00:00Z, toDate: 2021-02-15T00:00:00Z 1`] = `
Object {
  "active": 1,
  "closedWon": 0,
  "completeDownPayment": 0,
  "downPayment": 0,
  "inactive": 3,
  "tokenGenerated": 1,
}
`;

exports[`LeadsService listSummary should return the summary by considering only the tasks as well as winning tasks that were created against the lead in the provided interval i.e. fromDate : 2021-01-01T00:00:00Z, toDate: 2021-04-15T00:00:00Z 1`] = `
Object {
  "active": 1,
  "closedWon": 1,
  "completeDownPayment": 0,
  "downPayment": 0,
  "inactive": 3,
  "tokenGenerated": 0,
}
`;

exports[`LeadsService listSummary should return the summary by considering only the tasks as well as winning tasks that were created against the lead in the provided interval i.e. fromDate : 2021-02-15T00:00:00Z, toDate: 2021-03-31T00:00:00Z 1`] = `
Object {
  "active": 1,
  "closedWon": 0,
  "completeDownPayment": 0,
  "downPayment": 1,
  "inactive": 3,
  "tokenGenerated": 0,
}
`;

exports[`LeadsService listSummary should return the summary by considering only the tasks as well as winning tasks that were created against the lead in the provided interval i.e. fromDate : undefined, toDate: 2021-04-15T00:00:00Z 1`] = `
Object {
  "active": 3,
  "closedWon": 1,
  "completeDownPayment": 0,
  "downPayment": 0,
  "inactive": 1,
  "tokenGenerated": 0,
}
`;

exports[`LeadsService listSummary should return the summary by considering only the tasks as well as winning tasks that were created against the lead in the provided interval i.e. fromDate : undefined, toDate: undefined 1`] = `
Object {
  "active": 3,
  "closedWon": 1,
  "completeDownPayment": 0,
  "downPayment": 0,
  "inactive": 1,
  "tokenGenerated": 0,
}
`;
